service/src/main/java/bio/terra/workspace/service/iam/SamService.java

```diff
@@ -901,6 +901,36 @@ public String getOrCreatePetSaEmail(String projectId, AuthenticatedUserRequest u
     }
   }
 
+  /**
+   * Get a pet service account token for a given project and user request
+   *
+   * @param projectId - GCP project ID
+   * @param userRequest - original user request
+   * @return - pet service account token
+   */
+  public String getOrCreatePetSaToken(String projectId, AuthenticatedUserRequest userRequest) {
+    GoogleApi googleApi = samGoogleApi(userRequest.getRequiredToken());
+    try {
+      return googleApi.getPetServiceAccountToken(projectId, new ArrayList<>(SAM_OAUTH_SCOPES));
+    } catch (ApiException apiException) {
+      throw SamExceptionFactory.create(
+          "Error getting pet service account token from Sam", apiException);
+    }
+  }
+
+  /**
+   * Get an AuthenticatedUserRequest object for the pet SA of this user.
+   * @param projectId - GCP project ID
+   * @param userRequest - User's own AuthenticatedUserRequest. May contain null email and/or subjectID
+   * @return AuthenticatedUserRequest with the token replaced by the pet token
+   */
+  public AuthenticatedUserRequest getAuthenticatedPetRequest(String projectId,
+      AuthenticatedUserRequest userRequest) {
+    final String petToken = getOrCreatePetSaToken(projectId, userRequest);
+    return new AuthenticatedUserRequest(
+        userRequest.getEmail(), userRequest.getSubjectId(), Optional.of(petToken));
```

Comment:
Instead of `Optional`, should this fail if `petToken` is null?

Reply:
I don't know why that constructor parameter is optional and the other ones, which frequently aren't there, aren't. It's one thing I'd like to fix in that class.

Reply:
`Optional.of()` will throw if it's null. We could also use `.orElseThrow()`.